Add support for defs in svg files.

vedelaar opened the pull request:
Fixes #3847

The idea is that we create a `map<string, shared_ptr<shape>>` (defs_lookup_list) and fill it with the items found in the `<defs>` tag. The ID of the thing being the string in the map. Then when we encounter a `<use>` tag we check the map for the ID it references to, we clone the thing in the map (including the children of the thing) and make it the child of the use tag. 

X and Y information can be added to the `<use>` tag and that is stored as a "translate" in the transform string in the `<use>` shape. That way when transformations are applied the children of the use tag have the proper location.

t-paul commented:
Nice! I'm wondering about the test case image update. Looking at the SVG source, I think it should place the filled ellipsis not only in the initial position but also overlay in all the 4 arc examples.

I did some testing with
```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg xmlns:svg="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="20mm" height="2mm" viewBox="0 -1 20 2">
  <defs>
    <circle id="circle" cx="5" cy="0" r="1" />
    <rect id="rect" x="5" y="1" width="1" height="1" rx="0.1" />
  </defs>

  <circle cx="0" cy="0" r="1" />
  <rect id="rect" x="0" y="1" width="1" height="1" rx="0.1" />

  <use xlink:href="#circle" />
  <use xlink:href="#rect" />

  <use x="10" xlink:href="#circle" />
  <use x="10" xlink:href="#rect" />

  <g transform="translate(15 0)">
    <use xlink:href="#circle" />
    <use xlink:href="#rect" />
  </g>
</svg>                                                                                                                                                                                                             
```
And that looks good.

vedelaar commented:
I looked into the test case image. There was a problem in my code that prevented children 2 deep or more in a `<defs>` to have the correct parent and because of that were missing on some translate calls. I fixed that.

After fixing that the test case has 5 times the same image and the original test gets lost. I changed the svg file so that the original test is kept, and also added an extra test case to test my change better.

I rebased and force-pushed the commit to keep history clean, if you are interested in what I changed after your comment: https://github.com/openscad/openscad/compare/40e8fec..b844d7c

t-paul reviewed (approved):
Thanks!
